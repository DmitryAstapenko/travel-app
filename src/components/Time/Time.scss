.time {
  position: absolute;
  top: 0;
  right: 0;
  background: #4d4d4d5c;
  border-radius: 5px;
  background-color: #ffffff8c;
  & h2 {
    text-align: center;
  }
  & > * {
    margin: 0;
  }
}
