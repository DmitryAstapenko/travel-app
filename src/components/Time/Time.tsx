import React, { FC, useCallback, useEffect, useState } from "react";
import "./Time.scss";

interface ITime {
  date: string;
  month: string;
  day: string;
  hour: string;
  minute: string;
  second: string;
}

interface ITimeProps {
  timezone: string;
}

const Time: FC<ITimeProps> = ({ timezone }: ITimeProps) => {
  const addZero = (number: number) => `${number < 10 ? "0" : ""}${number}`;
  const clockTimeZone = useCallback(() => {
    const timeZoneValues = timezone.match(/^(?:UTC\+)(\d{2})/);
    const timezoneOffset = +(timeZoneValues ? timeZoneValues[1] : 0);
    const localTime = new Date();
    const ms =
      localTime.getTime() +
      localTime.getTimezoneOffset() * 60000 +
      timezoneOffset * 3600000;
    const timeOfZone = new Date(ms);

    const hour = addZero(timeOfZone.getHours());
    const minute = addZero(timeOfZone.getMinutes());
    const date = addZero(timeOfZone.getDate());
    const second = addZero(timeOfZone.getSeconds());
    const month = timeOfZone.toLocaleString("default", { month: "long" });
    const day = timeOfZone.toLocaleString("default", { weekday: "long" });
    return {
      hour,
      minute,
      date,
      month,
      day,
      second,
    };
  }, [timezone]);

  const [time, setTime] = useState<ITime>(clockTimeZone());

  useEffect(() => {
    let timeOut: NodeJS.Timeout;
    const getTime = () => {
      setTime(clockTimeZone());
      timeOut = setTimeout(getTime, 1000);
    };
    getTime();
    return () => {
      clearTimeout(timeOut);
    };
  }, [clockTimeZone]);
  const { hour, minute, date, month, day, second } = time;
  return (
    <div className="time">
      <h6>{`${date} ${month} ${day}`}</h6>
      <h2>{`${hour}:${minute}:${second}`}</h2>
    </div>
  );
};

export default Time;
