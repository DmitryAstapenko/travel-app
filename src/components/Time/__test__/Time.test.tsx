import React from "react";
import Time from "../Time";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

const setUp = (props) => shallow(<Time {...props} />);

describe("should render time", () => {
  let component;
  beforeEach(() => {
    component = setUp({ timezone: "UTC+03:00" });
  });
  it("should render time component", () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
