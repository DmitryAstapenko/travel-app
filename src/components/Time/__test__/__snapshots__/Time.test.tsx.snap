// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render time should render time component 1`] = `
<div
  className="time"
>
  <h6>
    17 M03 Wed
  </h6>
  <h2>
    15:54:44
  </h2>
</div>
`;
